Tests/Realm.Tests/Sync/FlexibleSyncTests.cs

```diff
@@ -1100,620 +1091,584 @@ public void SubscriptionSet_Enumerator_Reset()
+            Assert.That(sessionError.ErrorCode, Is.EqualTo(ErrorCode.WriteNotAllowed));
+        }
 
-                realm.Write(() =>
-                {
-                    ato.Int32Property = 15;
-                });
+        [Test, RequiresSynchronizationContext, RequiresBaas]
+        public async Task Integration_UpdateObjectNotMatchingSubscriptions_ShouldError()
+        {
+            var testGuid = Guid.NewGuid();
+            var realm = await GetFLXIntegrationRealmAsync();
 
-                var sessionError = await sessionErrorTask;
-                Assert.That(sessionError.ErrorCode, Is.EqualTo(ErrorCode.WriteNotAllowed));
+            realm.Subscriptions.Update(() =>
+            {
+                realm.Subscriptions.Add(realm.All<SyncAllTypesObject>().Where(o => o.GuidProperty == testGuid));
             });
-        }
 
-        [Test]
-        public void Integration_SubscriptionOnUnqueryableField_ShouldError()
-        {
-            SyncTestHelpers.RunBaasTestAsync(async () =>
+            var sessionErrorTask = SyncTestHelpers.WaitForSessionError<ClientResetException>(session =>
             {
-                var realm = await GetFLXIntegrationRealmAsync();
+                Assert.That(session.Path, Is.EqualTo(realm.Config.DatabasePath));
+            });
 
-                realm.Subscriptions.Update(() =>
+            var ato = realm.Write(() =>
+            {
+                return realm.Add(new SyncAllTypesObject
                 {
-                    realm.Subscriptions.Add(realm.All<SyncAllTypesObject>().Where(o => o.StringProperty == "foo"));
+                    GuidProperty = testGuid,
                 });
+            });
 
-                try
-                {
-                    await realm.Subscriptions.WaitForSynchronizationAsync();
-                    Assert.Fail("Expected an error to be thrown.");
-                }
-                catch (SubscriptionException ex)
-                {
-                    Assert.That(ex.Message, Does.Contain(nameof(SyncAllTypesObject.StringProperty)).And.Contains(nameof(SyncAllTypesObject)));
-                }
+            await WaitForUploadAsync(realm);
+
+            realm.Write(() =>
+            {
+                ato.GuidProperty = Guid.NewGuid();
+            });
 
-                Assert.That(realm.Subscriptions.State, Is.EqualTo(SubscriptionSetState.Error));
-                Assert.That(realm.Subscriptions.Error.Message, Does.Contain(nameof(SyncAllTypesObject.StringProperty)).And.Contains(nameof(SyncAllTypesObject)));
+            realm.Write(() =>
+            {
+                ato.Int32Property = 15;
             });
+
+            var sessionError = await sessionErrorTask;
+            Assert.That(sessionError.ErrorCode, Is.EqualTo(ErrorCode.WriteNotAllowed));
         }
 
-        [Test]
-        public void Integration_AfterAnError_CanRecover()
+        [Test, RequiresSynchronizationContext, RequiresBaas]
```

Comment:
Would it make sense to make these attributes hierarchical? I.e. rename `[RequiresSynchronizationContext]` to `[SynchronizationContextTest]` and have it inherit from `[TestAttribute]` and have `[BaasTest]` inherit from `[SynchronizationContextTest]`? It'll reduce the amount of attributes we have to apply on each test.

Reply:
I like the composability, because it allows us to mix `RequiresSynchronizationContext` and `RequiresBaas` both with each other and with `Test`, `TestCase`, and `Theory`.
Plus with my latest change `RequiresSynchronizationContext` becomes inherited and we only need have it explicitly once on the base `RealmTest` so it's implicitly picked up everywhere.
As for `RequiresBaas` I think it's better to keep it as a separate action attribute because this allows us to apply it to both fixtures and test cases.